File: src/Plotly.NET/Traces/ObjectAbstractions/Treemap.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open DynamicObj
open System
open System.Runtime.InteropServices


type TreemapRoot() =
    inherit DynamicObj()

    static member init(?Color: Color) =

        TreemapRoot() |> TreemapRoot.style (?Color = Color)

    static member style(?Color: Color) =
        fun (root: TreemapRoot) ->
            root
            |> DynObj.withProperty "color" Color

type TreemapLeaf() =
    inherit DynamicObj()

    static member init(?Opacity: float) =

        TreemapLeaf() |> TreemapLeaf.style (?Opacity = Opacity)

    static member style(?Opacity: float) =
        fun (leaf: TreemapLeaf) ->

            leaf
            |> DynObj.withOptionalProperty "opacity" Opacity


type TreemapTiling() =
    inherit DynamicObj()

    static member init
        (
            ?Packing: StyleParam.TreemapTilingPacking,
            ?SquarifyRatio: float,
            ?Flip: StyleParam.TilingFlip,
            ?Pad: float
        ) =

        TreemapTiling()
        |> TreemapTiling.style (?Packing = Packing, ?SquarifyRatio = SquarifyRatio, ?Flip = Flip, ?Pad = Pad)

    static member style
        (
            ?Packing: StyleParam.TreemapTilingPacking,
            ?SquarifyRatio: float,
            ?Flip: StyleParam.TilingFlip,
            ?Pad: float
        ) =
        fun (tiling: TreemapTiling) ->

            tiling
            |> DynObj.withOptionalPropertyBy "packing" Packing StyleParam.TreemapTilingPacking.convert
            |> DynObj.withOptionalProperty "squarifyRatio" SquarifyRatio
            |> DynObj.withOptionalPropertyBy "flip" Flip StyleParam.TilingFlip.convert
            |> DynObj.withOptionalProperty "pad" Pad

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Treemap.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.TreemapRoot InitTreemapRoot
        (
            Optional<Color> Color = default
        ) 
            =>
                Plotly.NET.TraceObjects.TreemapRoot.init(
                    Color: Color.ToOption()
                );
    public static Plotly.NET.TraceObjects.TreemapLeaf InitTreemapLeaf
        (
            Optional<double> Opacity = default
        ) 
            =>
                Plotly.NET.TraceObjects.TreemapLeaf.init(
                    Opacity: Opacity.ToOption()
                );
    public static Plotly.NET.TraceObjects.TreemapTiling InitTreemapTiling
        (
            Optional<StyleParam.TreemapTilingPacking> Packing = default,
            Optional<double> SquarifyRatio = default,
            Optional<StyleParam.TilingFlip> Flip = default,
            Optional<double> Pad = default
        ) 
            =>
                Plotly.NET.TraceObjects.TreemapTiling.init(
                    Packing: Packing.ToOption(),
                    SquarifyRatio: SquarifyRatio.ToOption(),
                    Flip: Flip.ToOption(),
                    Pad: Pad.ToOption()
                );
}

